File: backend/api/src/Wayel.Application/Features/OpsAuth/OpsRegions.cs
namespace Wayel.Application.Features.OpsAuth;

/// <summary>
/// Functional areas of the ops console. Assigned per user on invite; drives
/// navigation visibility and API authorization.
/// </summary>
public static class OpsRegions
{
    public const string Receiving = "receiving";
    public const string Collection = "collection";
    public const string Warehouse = "warehouse";
    public const string Platform = "platform";

    public static readonly IReadOnlySet<string> All =
        new HashSet<string>(StringComparer.OrdinalIgnoreCase)
        {
            Receiving,
            Collection,
            Warehouse,
            Platform,
        };

    public static IReadOnlyList<string> Normalize(IEnumerable<string>? values)
    {
        if (values is null)
        {
            return [];
        }

        var list = new List<string>();
        foreach (var raw in values)
        {
            var v = raw?.Trim().ToLowerInvariant();
            if (string.IsNullOrEmpty(v) || !All.Contains(v) || list.Contains(v, StringComparer.Ordinal))
            {
                continue;
            }

            list.Add(v);
        }

        return list;
    }

    /// <summary>
    /// When no explicit regions are stored, infer from the role (legacy users).
    /// </summary>
    public static IReadOnlyList<string> ResolveForRole(string role, IReadOnlyList<string>? storedRegions)
    {
        var explicitRegions = Normalize(storedRegions);
        if (explicitRegions.Count > 0)
        {
            return explicitRegions;
        }

        return role.Trim().ToLowerInvariant() switch
        {
            "receiver" => [Receiving],
            "collector" => [Collection],
            "finance" => [Receiving, Platform],
            "lead" => [Receiving, Collection, Warehouse, Platform],
            "clerk" => [Receiving, Warehouse, Collection],
            _ => [Receiving],
        };
    }
}

File: frontend/apps/we-yell-ops-dash/src/app/services/ops-regions.ts
/** Mirrors backend <see cref="OpsRegions"/> — functional areas of the ops console. */
export const OPS_REGION = {
  receiving: 'receiving',
  collection: 'collection',
  warehouse: 'warehouse',
  platform: 'platform',
} as const;

export type OpsRegion = (typeof OPS_REGION)[keyof typeof OPS_REGION];

export const OPS_REGION_LABELS: Record<OpsRegion, string> = {
  receiving: 'Receiving (South Africa)',
  collection: 'Collection (Eswatini)',
  warehouse: 'Warehouse',
  platform: 'Platform admin',
};

export const OPS_ROLE_REGION_PRESETS: Record<string, OpsRegion[]> = {
  lead: [
    OPS_REGION.receiving,
    OPS_REGION.collection,
    OPS_REGION.warehouse,
    OPS_REGION.platform,
  ],
  clerk: [OPS_REGION.receiving, OPS_REGION.warehouse, OPS_REGION.collection],
  finance: [OPS_REGION.receiving, OPS_REGION.platform],
  receiver: [OPS_REGION.receiving],
  collector: [OPS_REGION.collection],
};

export function normalizeOpsRegions(values: readonly string[] | undefined | null): OpsRegion[] {
  const allowed = new Set<string>(Object.values(OPS_REGION));
  const out: OpsRegion[] = [];
  for (const raw of values ?? []) {
    const v = raw?.trim().toLowerCase();
    if (!v || !allowed.has(v) || out.includes(v as OpsRegion)) {
      continue;
    }
    out.push(v as OpsRegion);
  }
  return out;
}

export function regionsForRole(role: string, stored?: readonly string[] | null): OpsRegion[] {
  const explicit = normalizeOpsRegions(stored);
  if (explicit.length > 0) {
    return explicit;
  }
  return OPS_ROLE_REGION_PRESETS[role.trim().toLowerCase()] ?? [OPS_REGION.receiving];
}

export function defaultHomePath(regions: readonly OpsRegion[]): string {
  if (regions.includes(OPS_REGION.collection) && !regions.includes(OPS_REGION.receiving)) {
    return '/ops/collection';
  }
  if (regions.includes(OPS_REGION.receiving)) {
    return '/ops/receiving/new';
  }
  if (regions.includes(OPS_REGION.collection)) {
    return '/ops/collection';
  }
  if (regions.includes(OPS_REGION.warehouse)) {
    return '/ops/warehouse';
  }
  return '/ops/receiving';
}
